agent/lib/kontena/workers/node_info_worker.rb

```diff
@@ -245,6 +264,93 @@ def calculate_container_time(container)
       0
     end
 
+    # @param [Array<Vmstat::Cpu>] prev_cpus
+    # @param [Array<Vmstat::Cpu>] current_cpu
+    # @return [Hash] { :num_cores, :system, :user, :idle }
+    def calculate_cpu_usage(prev_cpus, current_cpus)
+      result = {
+        num_cores: prev_cpus.size,
+        system: 0.0,
+        user: 0.0,
+        nice: 0.0,
+        idle: 0.0
+      }
+
+      prev_cpus.zip(current_cpus).map { |prev_cpu, current_cpu|
+        system_ticks = current_cpu.system - prev_cpu.system
+        user_ticks = current_cpu.user - prev_cpu.user
+        nice_ticks = current_cpu.nice = prev_cpu.nice
+        idle_ticks = current_cpu.idle - prev_cpu.idle
+
+        total_ticks = (system_ticks + user_ticks + nice_ticks + idle_ticks).to_f
+
+        {
+          system: (system_ticks / total_ticks) * 100.0,
+          user: (user_ticks / total_ticks) * 100.0,
+          nice: (nice_ticks / total_ticks) * 100.0,
+          idle: (idle_ticks / total_ticks) * 100.0
+        }
+      }.inject(result) { |memo, cpu_core|
+        memo[:system] += cpu_core[:system]
+        memo[:user] += cpu_core[:user]
+        memo[:nice] += cpu_core[:nice]
+        memo[:idle] += cpu_core[:idle]
+        memo
+      }
+    end
+
+    # @param [Array<Vmstat::NetworkInterface>] prev_interfaces
+    # @param [Array<Vmstat::NetworkInterface>] current_interfaces
+    # @param [Number] interval_seconds
+    # @return [Hash]
+    def calculate_network_traffic(prev_interfaces, current_interfaces, interval_seconds)
+      internal_interfaces = current_interfaces.select { |iface|
+        iface.name.to_s == "weave" or iface.name.to_s.start_with?("vethwe")
+      }
+
+      external_interfaces = current_interfaces.select { |iface|
+        iface.name.to_s == "docker0"
+      }
+
+      {
+        internal: calculate_interface_traffic(prev_interfaces, internal_interfaces, interval_seconds),
+        external: calculate_interface_traffic(prev_interfaces, external_interfaces, interval_seconds)
+      }
+    end
+
+    def calculate_interface_traffic(prev_interfaces, current_interfaces, interval_seconds)
+      results = {
+        interfaces: [],
+        rx_bytes: 0,
+        prev_rx_bytes: 0,
```

Comment:
It's very possible (immediately after a reboot) for e.g. the `weave` interfaces to be missing during `#initialize`, but appear before `publish_node_stats` first runs. OTOH, those newly created interfaces are going to start with their counters at zero, so I guess this all works out.